### ISSUE TITLE
Perhaps transmission-create may encounter some issues when creating a torrent for large files.

### ISSUE THREAD

**Issue by cosikng** [transmission/transmission#6839]
### What's the problem?
I've been experiencing some issues lately while creating torrents using Transmission. After creating the torrent with transmission-create and starting seeding, Transmission begins to verify the local files. However, it always gets stuck at around 99%. No matter how many times I retry 'Verify Local Data', it remains stuck. As a result, I had to recreate the torrent several times, leading me to delete the torrent already uploaded to the PT website. **However, the problem still persists.** This has caused me a lot of inconvenience. There are 78 files I want to seed, totaling approximately 30GB.

### What did I do to investigate the issue?
To investigate what was happening, I used transmission-create to create two torrents, both with a block size of 2048KB. I saved both of these torrents, distinguishing one of them by adding the '_new' suffix.
`transmission-create -s 2048 -o mytorrent.torrent /path/to/myfile`
Afterward, I wrote a simple Python script to extract the 'pieces' field from both of these torrents, which contains the hash values for all files.
```
import bencoder


def extract_pieces(torrent_file_path, output_file_path):
    try:
        with open(torrent_file_path, 'rb') as f:
            data = f.read()

        torrent_data = bencoder.bdecode(data)


        pieces_data = torrent_data[b'info'][b'pieces']


        with open(output_file_path, 'wb') as f:
            f.write(pieces_data)

        # print(pieces_data)

    except Exception as e:
        print(e)



torrent_file_path = 'mytorrent_new.torrent' 
output_file_path = 'mytorrent_new.bin'  
extract_pieces(torrent_file_path, output_file_path)
```
Then, I performed SHA256 checksums on these two extracted binary files and found them to be different. Furthermore, I used the following script to compare the extracted hash values byte by byte to identify the differing positions.
```
def compare_bytes(byte_str1, byte_str2):
    if len(byte_str1) != len(byte_str2):
        raise ValueError("Length not match")

    diff_indices = []
    for i in range(len(byte_str1)):
        if byte_str1[i] != byte_str2[i]:
            diff_indices.append(i)

    return diff_indices


def display_difference(byte_str1, byte_str2, diff_indices):
    if not diff_indices:
        print("No difference")
    else:
        print("Difference:")
        for idx in diff_indices:
            print(f"Pos {idx}: {byte_str1[idx]} -> {byte_str2[idx]}")


with open('mytorrent.bin', 'rb') as f:
    byte_str1 = f.read(-1)
with open('mytorrent_new.bin', 'rb') as f:
    byte_str2 = f.read(-1)

diff_indices = compare_bytes(byte_str1, byte_str2)
display_difference(byte_str1, byte_str2, diff_indices)
```
I found that approximately three hundred blocks' hash values were different, which is roughly the size of one file. This torrent has a total of about 17,000 blocks.

I speculate that this might be an error occurring at certain positions during the creation of the torrent file by transmission-create. I am currently unable to determine whether this is an internal error of the program or a disk read/write error. However, I believe that even if it's a disk read/write error, transmission-create should be able to handle it correctly.

Out of consideration for community rules, I feel it's inappropriate to share my torrent files. However, I'm more than willing to provide any additional information if needed. I appreciate any help and advice.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.5

**Comment by cosikng** [comment]
### Correction:
I apologize for the error in counting the differing hash values between the two torrent files. The actual number of discrepancies should be around 15 blocks. Based on further calculations, this aligns well with the number of erroneous blocks encountered during local data verification within Transmission.

**Comment by tearfur** [comment]
I appreciate your efforts to investigate this.

It's really hard for us to debug without any specifics though. I have some questions:

1. Does this only happen with `transmission-create`? Or other torrent creators as well?
2. Is there any public domain file where this can be reproduced?

**Comment by cosikng** [comment]
I am using an embedded Linux system (similar to a Raspberry Pi) as a seed server, running Transmission on it. I haven't tried other programs for creating torrents on it. I am still trying to reproduce the hash value error that occurs when creating a torrent. For example, I tried using the `dd` command to fully load my disk while creating a torrent with `transmission-create`, but I haven't found a consistent method to reproduce the issue. If it would be helpful, I can share the torrents that were created incorrectly and the ones that were created correctly (without tracker information and not distributed on DHT). If there is anything else you need me to provide, please let me know.

**Comment by Pentaphon** [comment]
> I have a hypothesis: could this issue be caused by `transmission-create` encountering very high disk read/write latency during torrent creation? After all, my device is connected to a relatively poor-performing mechanical hard drive.

No way to tell unless you provide a file for us or upgrade your hardware to SSD.

**Comment by tearfur** [comment]
> I have a hypothesis: could this issue be caused by `transmission-create` encountering very high disk read/write latency during torrent creation? After all, my device is connected to a relatively poor-performing mechanical hard drive.

I don't see how that can happen... AFAICT we only do synchronised read/write.

**Comment by Ercsion** [comment]
same issues for me, when i create a private torrent for 239.58 GB folder, the metadata file always contain some broken pieces, at least one pices.

```shell
# transmission-daemon -V
transmission-daemon 4.0.6 (38c164933e)

# transmission-create -V
transmission-create 4.0.6 (38c164933e)

# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 24.10
Release:        24.10
Codename:       oracular

# uname -a
Linux ubuntu2204 6.11.0-9-generic #9-Ubuntu SMP PREEMPT_DYNAMIC Mon Oct 14 13:19:59 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

# sudo lshw -class disk
  *-disk:0                  
       description: ATA Disk
       product: WDC WD120EDAZ-11
       vendor: Western Digital
       physical id: 0
       bus info: scsi@2:0.0.0
       logical name: /dev/sda
       version: 0A81
       serial: 8CKM****
       size: 10TiB (12TB)
       capabilities: gpt-1.00 partitioned partitioned:gpt
       configuration: ansiversion=5 guid=********-****-****-****-************ logicalsectorsize=512 sectorsize=4096

```

**Comment by Ercsion** [comment]
> I appreciate your efforts to investigate this.
> 
> It's really hard for us to debug without any specifics though. I have some questions:
> 
> 1. Does this only happen with `transmission-create`? Or other torrent creators as well?
> 2. Is there any public domain file where this can be reproduced?

I tryed [mktorrent](https://github.com/pobrn/mktorrent) on ubuntu x64 and [qBittorrent](https://www.fosshub.com/qBittorrent-old.html?dwl=qbittorrent_4.3.2_x64_setup.exe) on window10, create torrent for the same folder
qbittorrent works fine but mktorrent had the same issuse

**Comment by tearfur** [comment]
Hmm, just a wild guess. Continuing on the idea where there might be a problem with `stat()`, maybe it has something to do with calling `stat()` in quick succession?

**Comment by Ercsion** [comment]
> Hmm, just a wild guess. Continuing on the idea where there might be a problem with `stat()`, maybe it has something to do with calling `stat()` in quick succession?

`tr_verify_worker::verify_torrent()` in `verify.cc` do the same thing with `tr_metainfo_builder::blocking_make_checksums()` in `tmakemeta.cc`. the first one not only calculate sha1 hash piece by piece but also compare whith metainfo. 

I think dding a comparison with the correct metainfo to `make_checksums()` may help

**Comment by tearfur** [comment]
Nice work! The results are bewildering though...

They seem to suggest `transmission-create` modified a file, but we never opened a file in write mode.

Is it possible to find out which line triggered the change? It's not straightforward, but maybe we can add debug logs to `transmission-create` with precise timestamps, then `tail -f` the torrent files and add a timestamp to each line as well.

```bash
tail -f /path/to/tor/files | xargs -IL date +"[%Y-%m-%dT%T.%N] L"
```

Then we just compare the timestamps.

**Comment by tearfur** [comment]
> I don't think this is an issue with `transmission-create`, because the `tr_sys_file_read` function directly calls the `open`/`read` functions in read-only mode to read the file.
> Additionally, I encounter the same error with [mktorrent](https://github.com/pobrn/mktorrent).
> I believe this is a system-level issue, specifically with the ext4 file system.

I completely agree, but I'm stumped on how to begin singling out the problem for a bug report upstream, especially since this doesn't happen to me.

It'd be great if we can catch a pattern from `transmission-create`.
